tests/end_to_end/gcc-division/gcc-o2-out.c

```diff
@@ -0,0 +1,93 @@
+void test(s32 x, u32 y) {
+    s32 temp_hi;
+    s32 temp_hi_2;
+    s32 phi_a0;
+    s32 phi_a0_2;
+    s32 phi_a0_3;
+    s32 phi_a0_4;
+    s32 phi_a0_5;
+    s32 phi_a0_6;
+
+    func_00400090((u32) ((s32) (x + ((u32) x >> 0x1F)) >> 1));
+    func_00400090(x / 3);
+    phi_a0 = x;
+    if (x < 0) {
+        phi_a0 = x + 3;
+    }
+    func_00400090((u32) (phi_a0 >> 2));
+    func_00400090(x / 5);
+    func_00400090(x / 6);
+    func_00400090(x / 7);
+    phi_a0_2 = x;
+    if (x < 0) {
+        phi_a0_2 = x + 7;
+    }
+    func_00400090((u32) (phi_a0_2 >> 3));
+    func_00400090(x / 9);
+    func_00400090(x / 0xA);
+    func_00400090(x / 0x64);
+    func_00400090(x / 0xFF);
+    phi_a0_3 = x;
+    if (x < 0) {
+        phi_a0_3 = x + 0xFF;
+    }
+    func_00400090((u32) (phi_a0_3 >> 8));
+    func_00400090(x / 0x168);
+    func_00400090(x / 0x3E8);
+    func_00400090(x % 2);
+    func_00400090(x % 3);
+    phi_a0_4 = x;
+    if (x < 0) {
+        phi_a0_4 = x + 3;
+    }
+    func_00400090(x - ((phi_a0_4 >> 2) * 4));
+    func_00400090(x % 5);
+    func_00400090(x % 6);
+    func_00400090(x % 7);
+    phi_a0_5 = x;
+    if (x < 0) {
+        phi_a0_5 = x + 7;
+    }
+    func_00400090(x - ((phi_a0_5 >> 3) * 8));
+    func_00400090(x % 9);
+    func_00400090(x % 0xA);
+    func_00400090(x % 0x64);
+    func_00400090(x % 0xFF);
+    phi_a0_6 = x;
+    if (x < 0) {
+        phi_a0_6 = x + 0xFF;
+    }
+    func_00400090(x - ((phi_a0_6 >> 8) << 8));
+    func_00400090(x % 0x168);
+    func_00400090(x % 0x3E8);
+    func_00400090(y >> 1);
+    func_00400090(y / 3);
+    func_00400090(y >> 2);
+    func_00400090(y / 5);
+    func_00400090(y / 6);
+    temp_hi = y / 7;
+    func_00400090((u32) (temp_hi + ((u32) (y - temp_hi) >> 1)) >> 2);
```

Comment:
The `y / 7` codegen from GCC is super weird. I didn't think this was worth supporting.

Reply:
Makes sense. Still, it's a bit sad that we're outputting code that's actively incorrect, and will confuse people if/when it comes up (I assume it's not just 7 but some small proportion of numbers?). OTOH, it's nice that we're doing *something* in this case, because MULT_HI's are empirically very hard for people to parse, and it's not that hard to figure out you need to remove the weird sign bit correction thing. Ideally we'd emit a "partially detected division" comment when there's no sign bit correction maybe. But I understand if you feel that's too much effort.